### ISSUE TITLE
A non root task can end up in the barrier

### ISSUE THREAD

**Issue by mratsim** [mratsim/weave#51]
Completely flummoxed by that: https://travis-ci.com/mratsim/weave/jobs/267385458

Confirmed twice, somehow a non-root task can execute `sync` in there: https://github.com/mratsim/weave/blob/6b7b8c3984a83c49c7d6761bc61fa7c616116891/weave/parallel_tasks.nim#L166-L175

**Comment by mratsim** [comment]
More details:

It's only triggered with LazyFlowvar and not with the normal one. So this may be due to a different behaviour of alloca?

**Comment by mratsim** [comment]
This also happens on Windows and can be reproduced reliably on my machine by compiling `parallel_for.nim`

Then setting WEAVE_NUM_THREADS > 2 would not trigger an error but having WEAVE_NUM_THREADs = 2 would:

```

with printing of the parent, just before the assert
```Nim
Strided Nested loops
-------------------------
ptr 0000000000DD4400 --> [fn = nil,
parent = ptr 0000000000DD4200 --> [fn = nil,
parent = nil,
prev = nil,
next = nil,
start = 0,
cur = 0,
stop = 0,
stride = 0,
futures = nil,
batch = 0,
isLoop = false,
hasFuture = false,
futureSize = 0,
data = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
prev = nil,
next = nil,
start = 2,
cur = 5,
stop = 4,
stride = 1,
futures = nil,
batch = 1,
isLoop = true,
hasFuture = false,
futureSize = 0,
data = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

D:\Devel\Nim\weave\weave\parallel_for.nim(210) parallel_for
D:\Devel\Nim\weave\weave\parallel_for.nim(206) main4
D:\Devel\Nim\weave\weave\runtime.nim(178) exit
D:\Devel\Nim\weave\weave\instrumentation\contracts.nim(82) sync
D:\Devel\Nim\nim\lib\system\assertions.nim(27) failedAssertImpl
D:\Devel\Nim\nim\lib\system\assertions.nim(20) raiseAssert
D:\Devel\Nim\nim\lib\system\fatal.nim(39) sysFatal
Error: unhandled exception: D:\Devel\Nim\weave\weave\instrumentation\contracts.nim(82, 15) `
  localCtx.worker.currentTask.isRootTask()`
    Contract violated for pre-condition at runtime.nim:103

  localCtx.worker.currentTask.isRootTask()
    The following values are contrary to expectations:

  localCtx.worker.currentTask.isRootTask()  [Worker 0]
 [AssertionError]
```

So what seem to happen is that the main thread gets to the barrier when in the inner task.

The proper fix is to implement nestable barriers (#31 and #41)

**Comment by mratsim** [comment]
Not too sure what happens in Travis CI for Mac, it doesn't happen on Azure Pipelines.
This should never happens without nested parallel for.

Regarding nestable barriers, it seems impossible in the current design, see #31

**Comment by mratsim** [comment]
solved in #72
